app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java

```diff
@@ -2294,16 +2311,67 @@ public void exportData(Function<Entity, Void> handler) throws RegistryStorageExc
     }
 
     /**
-     * @see RegistryStorage#importData(io.apicurio.registry.storage.impexp.EntityInputStream)
+     * @see RegistryStorage#importData(io.apicurio.registry.storage.impexp.EntityInputStream, boolean, boolean)
      */
     @Override
     @Transactional
-    public void importData(EntityInputStream entities) throws RegistryStorageException {
-        handles.withHandleNoException( handle -> {
+    public void importData(EntityInputStream entities, boolean preserveGlobalId, boolean preserveContentId) throws RegistryStorageException {
+        Map<Long, Long> contentIdMapping = new HashMap<>();
+        // To handle issue, when we do not know new content id before importing the actual content
+        ArrayList<ArtifactVersionEntity> waitingForContent = new ArrayList<>();
+
+        handles.withHandleNoException(handle -> {
             Entity entity = null;
-            while ( (entity = entities.nextEntity()) != null ) {
-                if (entity != null) {
-                    importEntity(handle, entity);
+            while ((entity = entities.nextEntity()) != null) {
+                if (entity.getEntityType() == EntityType.ArtifactVersion) {
+                    ArtifactVersionEntity artifactVersionEntity = (ArtifactVersionEntity) entity;
+                    if (!preserveGlobalId) {
+                        ((ArtifactVersionEntity) entity).globalId = -1;
+                    }
+                    if (!preserveContentId) {
+                        if (!contentIdMapping.containsKey(artifactVersionEntity.contentId)) {
+                            // Add to the queue waiting for content imported
+                            waitingForContent.add(artifactVersionEntity);
+                            continue;
+                        }
+                        // Content of the artifact was already imported we can use the new contentId
+                        artifactVersionEntity.contentId = contentIdMapping.get(artifactVersionEntity.contentId);
+                    }
+                }
+                if (entity.getEntityType() == EntityType.Content) {
+                    ContentEntity contentEntity = (ContentEntity) entity;
+                    if (!preserveContentId) {
+                        // When we do not want to preserve contentId, the best solution to import content is create new one with the contentBytes
+                        // It makes sure there won't be any conflicts
+                        long newContentId = -1;
+                        if (contentEntity.artifactType != null) {
+                            newContentId = createOrUpdateContent(handle, contentEntity.artifactType, ContentHandle.create(contentEntity.contentBytes));
+                        } else {
+                            if(contentEntity.canonicalHash == null) throw new RegistryStorageException("There is not enough information about content. Artifact Type and CanonicalHash are both missing.");
```

Comment:
Can you format this a bit differently?  Just a style issue:

```java
if (contentEntity.canonicalHash == null) {
    throw new RegistryStorageException("There is not enough information about content. Artifact Type and CanonicalHash are both missing.");
}
```

Just for code style consistency.